pkg/build/api/validation/validation.go

```diff
@@ -36,12 +40,43 @@ func ValidateBuildUpdate(build *buildapi.Build, older *buildapi.Build) field.Err
 		allErrs = append(allErrs, field.Invalid(field.NewPath("status", "phase"), build.Status.Phase, "phase cannot be updated from a terminal state"))
 	}
 	if !kapi.Semantic.DeepEqual(build.Spec, older.Spec) {
-		allErrs = append(allErrs, field.Invalid(field.NewPath("spec"), "content of spec is not printed out, please refer to the \"details\"", "spec is immutable"))
+		diff, err := diffBuildSpec(build.Spec, older.Spec)
+		if err != nil {
+			glog.V(2).Infof("Error calculating build spec patch: %v", err)
+			diff = "[unknown]"
+		}
+		detail := fmt.Sprintf("spec is immutable, diff: %s", diff)
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec"), "content of spec is not printed out, please refer to the details", detail))
 	}
 
 	return allErrs
 }
 
+func diffBuildSpec(newer buildapi.BuildSpec, older buildapi.BuildSpec) (string, error) {
+	version := buildapi.SchemeGroupVersion
+	codec := kapi.Codecs.LegacyCodec(version)
+	newerObj := &buildapi.Build{Spec: newer}
+	olderObj := &buildapi.Build{Spec: older}
+
+	newerJSON, err := runtime.Encode(codec, newerObj)
+	if err != nil {
+		return "", fmt.Errorf("error encoding newer: %v", err)
+	}
+	olderJSON, err := runtime.Encode(codec, olderObj)
+	if err != nil {
+		return "", fmt.Errorf("error encoding older: %v", err)
+	}
+	versioned, err := kapi.Scheme.ConvertToVersion(newerObj, version.String())
+	if err != nil {
+		return "", fmt.Errorf("error converting newer to versioned: %v", err)
+	}
+	patch, err := strategicpatch.CreateTwoWayMergePatch(olderJSON, newerJSON, versioned)
```

Comment:
can `versioned` just be an empty build?  I'm pretty sure it just uses it navigate metadata for strategic merge patches.